File: fhc_2014/QR_C.cpp
// Facebook Hacker Cup 2014 Qualification Round
// C. Tennison

#include <algorithm>
#include <string>
#include <iostream>
#include <sstream>
#include <vector>
#include <map>
#include <stdio.h>

using namespace std;

typedef map<long double, long double> DD;
typedef vector<DD> DDVec;

long double solve(int K, long double P_sun, long double P_rain, long double P_init, long double P_up, long double P_win, long double P_down, long double P_lose)
{
	const long double zero = 0.0;
	const long double one = 1.0;

	long double total = 0.0;
	DD::const_iterator it;
	DDVec current(K+1);
	current[0][P_init] = 1.0;
	for (int set_no = 0; set_no < K*2; ++set_no) {
		DDVec next(K+1);
		for (int wins = 0; wins < K+1; ++wins) {
			DD &s = current[wins];
			int loses = set_no - wins;
			if (loses >= K) {
				continue;
			}
			if (wins >= K) {
				for (it = s.begin(); it != s.end(); ++it) {
					total += it->second;
				}
				continue;
			}
			for (it = s.begin(); it != s.end(); ++it) {
				long double sun = it->first;
				long double p = it->second;
				long double sunny = p * sun;
				long double rainy = p - sunny;
				long double sunny_win = sunny * P_sun;
				long double sunny_lose = sunny - sunny_win;
				long double sunny_up = sunny_win * P_win;
				long double sunny_down = sunny_lose * P_lose;
				long double rainy_win = rainy * P_rain;
				long double rainy_lose = rainy - rainy_win;
				long double rainy_up = rainy_win * P_win;
				long double rainy_down = rainy_lose * P_lose;
				next[wins+1][min(one,sun+P_up)] += sunny_up+rainy_up;
				next[wins+1][sun] += sunny_win-sunny_up+rainy_win-rainy_up;
				next[wins][max(zero,sun-P_down)] += sunny_down+rainy_down;
				next[wins][sun] += sunny_lose-sunny_down+rainy_lose-rainy_down;
			}
		}
		current = next;
	}

	return total;
}

int main(int argc, char *argv[])
{
	string s;
	getline(cin, s);
	int T = atoi(s.c_str());
	for (int t = 0; t < T; ++t) {
		getline(cin, s);
		stringstream ss(s);
		int K = 0;
		long double P_sun, P_rain, P_init, P_up, P_win, P_down, P_lose;
		ss >> K >> P_sun >> P_rain >> P_init >> P_up >> P_win >> P_down >> P_lose;
		double ans = solve(K, P_sun, P_rain, P_init, P_up, P_win, P_down, P_lose);
		printf("Case #%d: %.016f\n", t+1, ans);
	}

	return 0;
}

File: fhc_2014/QR_C.py
# Facebook Hacker Cup 2014 Qualification Round
# C. Tennison

import sys

def solve():
  total = 0.0
  args = sys.stdin.readline().rstrip().split(' ')
  K = int(args[0])
  (P_sun, P_rain, P_init, P_up, P_win, P_down, P_lose) = [float(x) for x in args[1:]]
  current = [{P_init:1.0}]
  for set_no in range(K*2):
    next = [{} for x in range(K+1)]
    for wins,s in enumerate(current):
      loses = set_no - wins
      if loses >= K:
        continue
      if wins >= K:
        for x in s:
          total += s[x]
        continue
      for sun in s:
        p = s[sun]
        sunny = p * sun
        rainy = p - sunny
        sunny_win = sunny * P_sun
        sunny_lose = sunny - sunny_win
        sunny_up = sunny_win * P_win
        sunny_down = sunny_lose * P_lose
        rainy_win = rainy * P_rain
        rainy_lose = rainy - rainy_win
        rainy_up = rainy_win * P_win
        rainy_down = rainy_lose * P_lose
        w = [[sunny_up+rainy_up,min(1.0,sun+P_up)],[sunny_win-sunny_up+rainy_win-rainy_up,sun]]
        l = [[sunny_down+rainy_down,max(0.0,sun-P_down)],[sunny_lose-sunny_down+rainy_lose-rainy_down,sun]]
        for x in w:
          if x[1] in next[wins+1]:
            next[wins+1][x[1]] += x[0]
          else:
            next[wins+1][x[1]] = x[0]
        for x in l:
          if x[1] in next[wins]:
            next[wins][x[1]] += x[0]
          else:
            next[wins][x[1]] = x[0]
    current = next
  return total

cases = int(sys.stdin.readline())
for case in range(cases):
  r = solve()
  print("Case #%d: %f" % (case+1, r))
